src/cmds/qstat.c

```diff
@@ -599,75 +599,83 @@ trunc_value(char *value, int len, int wide_len, int wide)
 static void
 prt_nodes(char *nodes, int no_newl)
 {
-	int  i, l;
+	int  i, len;
 	char linebuf[78];
-	char *rest;
-	char *token;
-	char *token_cp;
-	char *subtoken;
-	char *node_name;
-	char *chunk;
-	struct	sockaddr_in check_ip;
+	char *rest = NULL;
+	char *saveptr = NULL;
+ 	char *token = NULL;
+	char *token_cp = NULL;
+	char *subtoken = NULL;
+	char *node_name = NULL;
+	char *chunk = NULL;
+	struct sockaddr_in check_ip;
 	int ret = 0;
 
 	if ((nodes == NULL) || (*nodes == '\0'))
 		return;
 
 	i = 0;
-	rest = malloc(strlen(nodes) + 1);
-	if(rest)
-		strcpy(rest, nodes);
-	else
+	rest = strdup(nodes);
+	if (rest == NULL)
 		exit_qstat("out of memory");
-	while ((token = strtok_r(rest, "+", &rest))){
-		token_cp = (char*)malloc(strlen(token) + 1);
-		if(token_cp)
-			strcpy(token_cp, token);
-		else
+	/* The exec_host string has the format <host1>/<T1>*<P1>[+<host2>/<T2>*<P2>+... ].
+	 * We are using '+' delimiter to find each <host1>/<T1>*<P1> string.
+	 */
+	token = strtok_r(rest, "+", &saveptr);
```

Comment:
Thank you for adding the comment. Why are you using the reentrant version of strtok?

Reply:
I find using the reentrant version of strtok() best at all times.  It saves you from doing silly things like calling strtok() to break on fields, and then calling strtok() again to break into subfields.  There is no harm in using it.

Reply:
Ahh... thanks @bhroam. I overlooked the fact there is another call to strtok here.